### PR TITLE
[CELEBORN-1056] Introduce Rest API of listing dynamic configuration

### PR DESCRIPTION
**Pull request by SteNicholas**
### What changes were proposed in this pull request?

Introduce Rest API of listing dynamic configuration `/listDynamicConfigs` to list the dynamic configs. The result of `/listDynamicConfigs` is as follows:

```
=========================== Dynamic Configuration ============================
celeborn.master.ha.ratis.raft.server.snapshot.auto.trigger.threshold          100000
celeborn.worker.flusher.buffer.size                                           64k
=========================== SYSTEM ============================
celeborn.master.ha.ratis.raft.server.snapshot.auto.trigger.threshold          200000
celeborn.worker.flusher.buffer.size                                           128k
=========================== TENANT ============================
=========================== Tenant: tenantId1 ============================
celeborn.master.ha.ratis.raft.server.snapshot.auto.trigger.threshold          300000
celeborn.worker.flusher.buffer.size                                           256k
=========================== TENANT_USER ============================
=========================== Tenant: tenantId1, Name: user1 ============================
celeborn.master.ha.ratis.raft.server.snapshot.auto.trigger.threshold          400000
celeborn.worker.flusher.buffer.size                                           512k
```

### Why are the changes needed?

Celeborn supports dynamic configuration with `ConfigService` at present. It's recommend to introduce Rest API of dynamic configuration management.

### Does this PR introduce _any_ user-facing change?

- Introduce Rest API of listing dynamic configuration: `/listDynamicConfigs?level=[system|tenant|tenant_user]&tenant=tenantId1&name=user1`.

### How was this patch tested?

- `HttpUtilsSuite#CELEBORN-1056: Introduce Rest API of listing dynamic configuration`

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
`Introduce Rest API of listing dynamic configuration management `? Since we may support setting restful API later

File: service/src/main/scala/org/apache/celeborn/server/common/http/HttpEndpoint.scala [hunk 1]
```diff
@@ -39,6 +42,22 @@ case object Conf extends HttpEndpoint {
     service.getConf
 }
 
+case object DynamicConfig extends HttpEndpoint {
```
Comment (AngersZhuuuu) [inline service/src/main/scala/org/apache/celeborn/server/common/http/HttpEndpoint.scala:45]: ListDynamicConfig?

File: service/src/main/scala/org/apache/celeborn/server/common/http/HttpEndpoint.scala [hunk 2]
```diff
@@ -39,6 +42,22 @@ case object Conf extends HttpEndpoint {
...
+  override def path: String = "/dynamicConfig"
```
Comment (AngersZhuuuu) [inline service/src/main/scala/org/apache/celeborn/server/common/http/HttpEndpoint.scala:46]: ditto

File: service/src/main/scala/org/apache/celeborn/server/common/HttpService.scala [hunk 3]
```diff
@@ -27,15 +33,77 @@ abstract class HttpService extends Service with Logging {
...
+      level: String,
+      tenant: String,
+      name: String): String = {
+    if (configService == null) {
+      s"Dynamic configuration is disabled. Please check whether to config `${CelebornConf.DYNAMIC_CONFIG_STORE_BACKEND.key}`."
+    } else {
+      val sb = new StringBuilder
+      sb.append("=========================== Dynamic Configuration ============================\n")
+      level.split(",").filter(_.nonEmpty).foreach {
+        configLevel =>
```
Comment (AngersZhuuuu) [inline service/src/main/scala/org/apache/celeborn/server/common/HttpService.scala:55]:
```
level.split(",").filter(_.nonEmpty).foreach { configLevel =>
   ....
}
```

File: service/src/main/scala/org/apache/celeborn/server/common/HttpService.scala [hunk 4]
```diff
@@ -27,15 +33,77 @@ abstract class HttpService extends Service with Logging {
...
+          {
```
Comment (AngersZhuuuu) [inline service/src/main/scala/org/apache/celeborn/server/common/HttpService.scala:56]: Can remove this `{}` pair

File: service/src/main/scala/org/apache/celeborn/server/common/HttpService.scala [hunk 5]
```diff
@@ -27,15 +33,77 @@ abstract class HttpService extends Service with Logging {
...
+            sb.append(
+              s"=========================== Level: $configLevel ============================\n")
+            if (ConfigLevel.SYSTEM.name().equals(configLevel)) {
+              sb.append(configs(configService.getSystemConfigFromCache.getConfigs))
+            } else {
+              val tenantConfigs =
+                if (tenant.isEmpty) configService.listRawTenantConfigsFromCache().asScala
+                else tenant.split(",").map { tenantId =>
+                  configService.getRawTenantConfigFromCache(tenantId)
+                }.toList
```
Comment (AngersZhuuuu) [inline service/src/main/scala/org/apache/celeborn/server/common/HttpService.scala:66]:
```
if (tenant.isEmpty) {
    configService.listRawTenantConfigsFromCache().asScala
} else {
    tenant.split(",").map { tenantId =>  configService.getRawTenantConfigFromCache(tenantId) }.toList
}
```

File: service/src/main/scala/org/apache/celeborn/server/common/http/HttpEndpoint.scala [hunk 6]
```diff
@@ -39,6 +42,22 @@ case object Conf extends HttpEndpoint {
...
+    s"The parameter level specifies the config levels of dynamic configs, which value is separated by commas. " +
+    s"The parameter tenant specifies the tenant ids of ${ConfigLevel.TENANT.name()} or ${ConfigLevel.TENANT_USER.name()} level, which value is separated by commas. " +
+    s"The parameter name specifies the user names of ${ConfigLevel.TENANT_USER.name()} level, which value is separated by commas."
+
+  override def handle(service: HttpService, parameters: Map[String, String]): String =
+    service.getDynamicConfig(
+      parameters.getOrElse("LEVEL", ConfigLevel.values().mkString(","))
+        .toUpperCase(Locale.ROOT).trim,
+      parameters.getOrElse("TENANT", "").trim,
+      parameters.getOrElse("NAME", "").trim)
```
Comment (AngersZhuuuu) [inline service/src/main/scala/org/apache/celeborn/server/common/http/HttpEndpoint.scala:58]: I think
1. If set level with system level, we only show system level config
2. If set level with tenant level, without tenant id, should list all tenant level config
3. If set level with tenant level with tenant id, should only show this tenant id 's config
4. If set level with tenant user level, without setting id and name, should show all user level
5. If set level with tenant user level, with setting tenant id, should this tenant id's all user level config
6. If set level with tenant user level, with setting id and name, should show only this user's user level config
7. If not set level, but set tenant id or name, should follow above each level's behavior.

**Review by AngersZhuuuu — COMMENTED** [review]
Also better provide a real restful result's string in desc?

**Comment by SteNicholas** [comment]
@AngersZhuuuu, thanks for detailed review. I have addressed above comments and updated the title and description of this pull request. PTAL.

**Comment by SteNicholas** [comment]
@AngersZhuuuu, thanks for detailed review. I have addressed above comments and updated the title and description of this pull request. PTAL. cc @RexXiong.

File: service/src/main/scala/org/apache/celeborn/server/common/HttpService.scala [hunk 7]
```diff
@@ -27,15 +33,107 @@ abstract class HttpService extends Service with Logging {
...
+        }.toList
+      }
+    tenantConfigs.sortBy(_.getTenantId).map { tenantConfig =>
+      s"""
+         |=========================== Tenant: ${tenantConfig.getTenantId} ============================
+         |${configs(tenantConfig.getConfigs)}""".stripMargin
+    }.mkString("\n")
+  }
+
+  private def tenantUserConfigs(tenant: String, name: String): String = {
```
Comment (RexXiong) [inline service/src/main/scala/org/apache/celeborn/server/common/HttpService.scala:92]: IMO and from the parameter of tenantUserConfigs, we should only support single tenant/single tenant with username to simplify and avoid the combinations of tenantId and usernames, if tenant/name are both empty we list all configs by config_level

**Review by RexXiong — APPROVED** [review]
LGTM, thanks。

**Comment by SteNicholas** [comment]
Merge to main(v0.5.0).

**Pull request closed without merging**